### PR TITLE
Fix Issue 22054 - Referencing a fwd-declared field results in many error messages

### PR DESCRIPTION
**Pull request by RazvanN7**

### CODE REVIEW

**Comment by MoonlightSentinel** [comment]
Note that your branch is missing #12671. That PR is related because it fixes `ClassDeclaration.search` (and others) to adhere to `IgnoreErrors` used during symbol resolution.

This patch suffices for current master and produces a somewhat better error message:

```patch
diff --git a/src/dmd/dclass.d b/src/dmd/dclass.d
index 9c5f0da2f..f9b4aba37 100644
--- a/src/dmd/dclass.d
+++ b/src/dmd/dclass.d
@@ -498,7 +498,7 @@ extern (C++) class ClassDeclaration : AggregateDeclaration
         if (!members || !symtab) // opaque or addMember is not yet done
         {
             // .stringof is always defined (but may be hidden by some other symbol)
-            if (ident != Id.stringof && !(flags & IgnoreErrors))
+            if (ident != Id.stringof && !(flags & IgnoreErrors) && semanticRun < PASS.semanticdone)
                 error("is forward referenced when looking for `%s`", ident.toChars());
             //*(char*)0=0;
             return null;
```

```
Error: no property `what` for type `test22054.exception`
```

**Comment by MoonlightSentinel** [comment]
Also would be good to test the behavior for opaque `struct`s and `enum`s (the latter won't work with your current patch - unless it's already handled another way)

**Comment by RazvanN7** [comment]
@MoonlightSentinel Your patch is superior, so I implemented that, however, I added a supplemental error for aggregated types.

File: src/dmd/typesem.d [hunk 1]
```diff
@@ -2387,8 +2387,12 @@ Expression getProperty(Type t, Scope* scope_, const ref Loc loc, Identifier iden
                     error(loc, "no property `%s` for type `%s`", ident.toChars(), mt.toPrettyChars(true));
                     if (auto dsym = mt.toDsymbol(scope_))
                         if (auto sym = dsym.isAggregateDeclaration())
+                        {
                             if (auto fd = search_function(sym, Id.opDispatch))
                                 errorSupplemental(loc, "potentially malformed `opDispatch`. Use an explicit instantiation to get a better error message");
+                            else if (!sym.members)
+                                errorSupplemental(loc, "`%s` `%s` is forward declared and does not have a body.", sym.kind, mt.toPrettyChars(true));
```
Comment (MoonlightSentinel) [inline src/dmd/typesem.d:2394]: D does not support forward declarations, this should probably refer to an opaque type.
Also, can this error message be triggered for an actual forward reference error?

Comment (RazvanN7) [reply src/dmd/typesem.d]: > D does not support forward declarations, this should probably refer to an opaque type.

changed.

> Also, can this error message be triggered for an actual forward reference error?

Sorry, I don't understand what you mean.

Comment (MoonlightSentinel) [reply src/dmd/typesem.d]: `symtab` might be `null` if this error occurs early into in the semantic of `exception`. But i think it's currently impossible to write code that could trigger such scenarios. (e.g. base classes/interface for opaque classes are rejected by the parser)

Comment (RazvanN7) [reply src/dmd/typesem.d]: Typically, forward reference errors are issued before `noMember` is called.

**Comment by thewilsonator** [comment]
Is this good to go?

**Comment by RazvanN7** [comment]
I would say so.

File: src/dmd/typesem.d [hunk 2]
```diff
@@ -2387,8 +2387,12 @@ Expression getProperty(Type t, Scope* scope_, const ref Loc loc, Identifier iden
                     error(loc, "no property `%s` for type `%s`", ident.toChars(), mt.toPrettyChars(true));
                     if (auto dsym = mt.toDsymbol(scope_))
                         if (auto sym = dsym.isAggregateDeclaration())
+                        {
                             if (auto fd = search_function(sym, Id.opDispatch))
                                 errorSupplemental(loc, "potentially malformed `opDispatch`. Use an explicit instantiation to get a better error message");
+                            else if (!sym.members)
+                                errorSupplemental(loc, "`%s %s` is opaque, therefore it does not have a body.", sym.kind, mt.toPrettyChars(true));
```
Comment (MoonlightSentinel) [inline src/dmd/typesem.d:2394]: - `does not have a body` is redundant
- the hint should point to the type declaration, not it's use.

```suggestion
                                errorSupplemental(sym.loc, "`%s %s` is opaque and has no members", sym.kind, mt.toPrettyChars(true));
```

**Review by MoonlightSentinel — COMMENTED** [review]
Otherwise LGTM

**Comment by RazvanN7** [comment]
@MoonlightSentinel done.

**Review by MoonlightSentinel — APPROVED** [review]

**Pull request merged**